### DISCUSSION TITLE
Scheduling the number of local gradients aggregated at each step

### DISCUSSION THREAD

**Discussion by unfazing** [horovod/horovod#2786] [Q&A]
**Environment:**
I am using a docker image, **horovod/horovod:0.20.0-tf2.3.0-torch1.6.0-mxnet1.6.0.post0-py3.7-cuda10.1**, pulled from docker hub on machines running Ubuntu 20.04

**Checklist:**
1. Did you search issues to find if somebody asked this question before? yes
2. If your question is about hang, did you read [this doc] (https://github.com/horovod/horovod/blob/master/docs/running.rst)? NIL
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)? NIL
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? yes

**Your question:**
Hello again.

I am looking for an implementation of [adaptive periodic averaging SGD](https://arxiv.org/pdf/2007.06134.pdf). Correct me if I am wrong, but the local gradient aggregation function (which is implemented by adding the backwards_passes_per_step argument in the optimizer), is the same as periodic averaging SGD. It is synchronous and reduces communication overhead by the number of times that the local gradient is aggregated.

My question is if there is a way to schedule this parameter such that it is a function of the current learning rate/training loss?

Thank you.

**Comment by tgaddair** [comment]
Hey @unfazing, are you using TensorFlow or PyTorch? For PyTorch it should be fairly straightforward by modifying the `backward_passes_per_step` value programmatically.  It should also be doable for TensorFlow, but will require use of a tf.Variable we modify.

**Comment by unfazing** [comment]
I am using Tensorflow. Thank you for your response I will look into it.

**Comment by unfazing** [comment]
Hello @tgaddair, can I trouble you to provide me with a simple example in Tensorflow using the tf.Variable. 

I eventually wish to implement the `backward_pass_per_step` value to be proportional to `(current epoch training loss/previous epoch training loss) ** 0.5`; and perhaps scale with a scheduled learning rate as well.

**Comment by unfazing** [comment]
Hi @tgaddair, I am working on a Tensorflow training loop similar to the one seen [here](https://github.com/horovod/horovod/blob/master/examples/tensorflow2/tensorflow2_mnist.py). Can I clarify a few questions:
1. Is it correct to say broadcasting of variables only needs to be done for the first batch of the first epoch? (Or does it need to be broadcasted for the first batch of every epoch?)
2. The optimizer does not need to be wrapped by the hvd.DistributedOptimizer function?
3. If 2 is true, then how will local gradients be aggregated using Tensorflow training loops?

**Comment by tgaddair** [comment]
Hey @unfazing,

1. For broadcasting, we normally only do it during the first batch, because afterwards we assume all workers are in sync, which should be mostly true except in some rare cases of numerical non-determinism. So I would only do it once for the whole training job.
2. For `tf.function` training, we use the `DistributedGradientTape` instead of `DistributedOptimizer`, yes.
3. When using the gradient tape, this problem should become a lot easier. You just need to store off the grads each batch, aggregate them, and only wrap in the distributed gradient tape / apply once every k batches.

So mostly just this part would change:

```
tape = hvd.DistributedGradientTape(tape)
grads = tape.gradient(loss_value, mnist_model.trainable_variables)
opt.apply_gradients(zip(grads, mnist_model.trainable_variables))
```

**Comment by unfazing** [comment]
Thank you for your response @tgaddair. 

The local gradient aggregation is described as such on the Uber Eng blog:

> It works by accumulating gradient updates to the model locally in GPU memory, summing the latest updates with the previously accumulated updates since the last round of communication. After a configured number of mini-batches N, Horovod performs an allreduce to average gradients across workers over the network, then applies the updates to the model.

I’m not sure if my understanding is clear, please correct me if I’m wrong: 
1) Each local gradient update involves updating the local model variables. This can be implemented with `tf.GradientTape`, and calling `grads = tape.gradient(loss_value, mnist_model.trainable_variables)
opt.apply_gradients(zip(grads, mnist_model.trainable_variables))`

2) After N mini-batches, wrap the tape with `hvd.DistributedGradientTape` and apply gradients. The AllReduce function across workers is called during this step

3) What does ‘ summing the latest updates with the previously accumulated updates since the last round of communication.’ mean? Why is it necessary to do this?

*edit:
I just came across issue #2710, and saw that for the local gradient aggregation, the local model variables are not updated, effectively increasing batch size by k times. I suppose if I want to update the local model variables with each step, 1) holds true?

**Comment by tgaddair** [comment]
Hey @unfazing, as you said, typically we would not update the local model until we aggregate all the gradients across all the workers.  This is to prevent the model replicas from diverging and exploring completely different areas of the parameter space.

One way to fix this is to apply the approach from [Local SGD](https://arxiv.org/abs/1808.07217), which averages the model params across workers every k steps, not the gradients (see #2699).

cc @richardliaw who is investigating this for PyTorch in case you would like to compare notes.

**Comment by richardliaw** [comment]
Yeah @unfazing I think the #2699 branch that @tgaddair is what you're looking for (for pytorch, which is probably easier to work with than TF).

**Comment by unfazing** [comment]
Ah okay, good to know that it is in the works! I am looking forward to the Tensorflow implementation. Thanks for your help @tgaddair and @richardliaw.

**Comment by unfazing** [comment]
Hi was looking into this issue again and was wondering if Local SGD (workers compute gradients and update local model params. After k updates, model params are averaged across all workers) could be implemented as such:

```
@tf.function
def train_step(x, y):
    with tf.GradientTape() as tape:
        logits = model(x, training = True)
        loss_value = loss_fn(y, logits)

    grads = tape.gradient(loss_value, model.trainable_weights)
    optimizer.apply_gradients(zip(grads, model.trainable_weights)

    return loss_value

for epoch in range(epochs):
    for step, (x_batch_train, y_batch_train) in enumerate(train_dataset):
        if (step+1) % k == 0:
            for n in range(len(model.variables)):
                avg_v = hvd.allreduce(model.variables[n])
                model.variables[n].assign(avg_v)
        loss_value = train_step(x_batch_train, y_batch_train)

```
Did not include code for first batch broadcast and learning rate warmup for simplicity.

**Comment by tgaddair** [comment]
Hey @unfazing, yes I think that's the right idea. The one change I would suggest is that we can parallelize the `hvd.allreduce(model.variables[n])` by defining the ops within a graph similar to this function: https://github.com/horovod/horovod/blob/master/horovod/tensorflow/functions.py#L47
